cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionService.java

```diff
@@ -0,0 +1,314 @@
+        responder.sendString(HttpResponseStatus.INTERNAL_SERVER_ERROR,
+                             "Failed to encode transaction: " + e.getMessage());
+      }
+    }
+  }
+
+  /**
+   * A private class for handling the {@link Transaction} lifecycle for a job.
+   */
+  private final class JobTransaction {
+    private final Integer jobId;
+    private final Set<Integer> stageIds;
+    private final boolean ownTransaction;
+    private volatile Optional<Transaction> transaction;
+
+    JobTransaction(Integer jobId, Set<Integer> stageIds, @Nullable Transaction transaction) {
+      this.jobId = jobId;
+      this.stageIds = ImmutableSet.copyOf(stageIds);
+      this.ownTransaction = transaction == null;
+      this.transaction = transaction == null ? null : Optional.of(transaction);
+    }
+
+    Integer getJobId() {
+      return jobId;
+    }
+
+    Set<Integer> getStageIds() {
+      return stageIds;
+    }
+
+    /**
+     * Returns the {@link Transaction} associated with the job. If transaction hasn't been started, a new long
+     * transaction will be started.
+     *
+     * @return the job's {@link Transaction} or {@code null} if it failed to start transaction for the job.
+     */
+    @Nullable
+    public Transaction getTransaction() {
+      Optional<Transaction> tx = transaction;
+      if (tx == null) {
+        // double-checked locking
+        synchronized (this) {
+          tx = transaction;
+          if (tx == null) {
+            try {
+              tx = transaction = Optional.of(txClient.startLong());
+            } catch (Throwable t) {
+              LOG.error("Failed to start transaction for job {}", jobId, t);
+              tx = transaction = Optional.absent();
+            }
+          }
+        }
+      }
+      return tx.orNull();
+    }
+
+    /**
+     * Completes the job {@link Transaction} by either committing or invalidating the transaction, based on the
+     * job result.
+     *
+     * @param succeeded {@code true} if the job execution completed successfully.
+     */
+    public void completed(boolean succeeded) {
+      // If this service doesn't own the transaction, then nothing to do. It is for the explicit transaction case.
+      if (!ownTransaction) {
+        return;
+      }
+
+      try {
+        Optional<Transaction> tx = transaction;
+        if (tx == null || !tx.isPresent()) {
+          // No transaction was started for the job, hence nothing to do.
+          return;
+        }
+
+        Transaction jobTx = tx.get();
+        if (succeeded) {
+          LOG.debug("Committing transaction for job {}", jobId);
+          if (!txClient.commit(jobTx)) {
+            // If failed to commit (which it shouldn't since there is no conflict detection), invalidate the tx
+            Transactions.invalidateQuietly(txClient, jobTx);
+          }
+        } else {
+          LOG.debug("Aborting transaction for job {}", jobId);
+          txClient.invalidate(jobTx.getWritePointer());
+        }
```

Comment:
some Datasets (tx-aware) have a afterCommit() method. Will that ever be called in Spark?